### PR TITLE
Support environment variables with empty value

### LINKED ISSUE #34446: 'ProcessStartInfo.EnvironmentVariables ["FOO"] = null' launches with empty environment variable
Repro:

```csharp
class Program
{
    static void Main(string[] args)
    {
        var process = new System.Diagnostics.Process();
        process.StartInfo.FileName = "env";
        process.StartInfo.UseShellExecute = false;
        process.StartInfo.EnvironmentVariables ["FOO"] = null;
        process.Start();
        process.WaitForExit ();
    }
}
```

Execute:

```shell
$ dotnet run | grep FOO
FOO=
```

Try with mono:

```shell
csc *.cs /nologo && mono *.exe | grep FOO
[no output]
```

Allowing setting an environment variable to null to remove it would match the behavior of the Environment class, where you're supposed to `Environment.SetEnvironmentVariable ("FOO", null);` to delete that variable.

### LINKED ISSUE #50554: Environment.SetEnvironmentVariable does not support setting empty values
### Description

`Environment.SetEnvironmentVariable` internally normalizes the `value` argument from empty string to `null`:
https://github.com/dotnet/runtime/blob/4523fbfc78b78a797d662b7952ac674a3616ed7d/src/libraries/System.Private.CoreLib/src/System/Environment.cs#L233-L237

This is preventing the callers of this API to set variables to an empty value, which is otherwise supported on all platforms.

Consider the following code on Windows:

```C#
using System;
using System.Runtime.InteropServices;

namespace ConsoleApp
{
    class Program
    {
        [DllImport("kernel32", CharSet = CharSet.Unicode)]
        internal static extern bool SetEnvironmentVariable(string lpName, string? lpValue);

        static void Main(string[] args)
        {
            SetEnvironmentVariable("_MYVAR", "");
            Console.WriteLine(Environment.GetEnvironmentVariables().Contains("_MYVAR"));

            Environment.SetEnvironmentVariable("_MYVAR", "");
            Console.WriteLine(Environment.GetEnvironmentVariables().Contains("_MYVAR"));
        }
    }
}
```

It prints `True False` as the P/Invoke works as expected and the BCL call deletes the variable, same as if `Environment.SetEnvironmentVariable("_MYVAR", null)` was called.

This program is expected to print `True True`.

### Configuration

.NET Core 5.0.201 on Windows 10

### Regression?

Not a regression, this behavior was inherited from .NET Framework.

### Other information

I am proposing to make a breaking change to the API and remove the `""` -> `null` normalization. If this is not possible, please consider adding a new API that would allow the program to set environment variables to empty values.

Note that `ProcessStartupInfo.EnvironmentVariables` supports empty values and has actually the opposite problem of interpreting `null` as `""` (https://github.com/dotnet/runtime/issues/34446).

This issue was found when moving an out-of-proc invocation of a tool to in-process and trying to keep the same logic of passing environment variables (previously via `ProcessStartupInfo`) to the current process with `Environment`.

### PR DESCRIPTION
**Pull request by jkotas**
`Environment.SetEnvironment` API did not make distinction between null and empty string value. It treated both as request to delete the environment variable. It made it impossible to create an environment variable with empty string value. This change updates `Environment.SetEnvironment` to treat empty string value as empty string value (null value is still treated as request to delete the environment variable).

Also, this change updates ProcessStartInfo.Environment to treat null vs. empty string the same way as `Environment.SetEnvironment` for consistency.
   
Fixes #50554
Fixes #34446

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.SetEnvironmentVariable.cs [hunk 1]
```diff
@@ -262,14 +262,27 @@ public void DeleteEnvironmentVariable(EnvironmentVariableTarget target)
...
+            const string value = "false";
 
+            ExecuteAgainstTarget(target,
+            () =>
+            {
+                // First set the value to something and then ensure that it can be deleted.
                 Environment.SetEnvironmentVariable(varName, value);
-                Environment.SetEnvironmentVariable(varName, NullString);
+                Environment.SetEnvironmentVariable(varName, null);
                 Assert.Null(Environment.GetEnvironmentVariable(varName));
```
Comment (akoeplinger) [inline src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.SetEnvironmentVariable.cs:286]: should we also add a test to make sure `Environment.GetEnvironmentVariables()` does not contain the variable?

**Comment by akoeplinger** [comment]
Also fixes https://github.com/dotnet/runtime/issues/34446 I assume?

File: src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.SetEnvironmentVariable.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Runtime/tests/System.Runtime.Extensions.Tests/System/Environment.SetEnvironmentVariable.cs:286]: Fixed

**Comment by jkotas** [comment]
> Also fixes https://github.com/dotnet/runtime/issues/34446 I assume?

It does now - worth fixing for consistency.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 2]
```diff
@@ -612,10 +612,16 @@ private static string[] CreateEnvp(ProcessStartInfo psi)
         {
             var envp = new string[psi.Environment.Count];
             int index = 0;
-            foreach (var pair in psi.Environment)
+            foreach (KeyValuePair<string, string?> pair in psi.Environment)
             {
-                envp[index++] = pair.Key + "=" + pair.Value;
+                // Ignore null values for consistency with Environment.SetEnvironmentVariable
+                if (pair.Value != null)
```
Comment (jkotas) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:618]: I am waiting to see first whether there are existing tests that hit this case, and then add as necessary.

**Review by stephentoub — APPROVED** [review]

**Comment by jkotas** [comment]
Breaking change issue created: https://github.com/dotnet/docs/issues/41471

**Comment by jkotas** [comment]
/ba-g Known issue #103642 that has been fixed and closed already

**Pull request merged**